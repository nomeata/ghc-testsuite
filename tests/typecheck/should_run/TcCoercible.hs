{-# LANGUAGE RoleAnnotations #-}

import GHC.Prim (coerce)
import GHC.Types (Coercible)

newtype Age = Age Int deriving Show
newtype Foo = Foo Age deriving Show
newtype Bar = Bar Age deriving Show
newtype Baz = Baz Bar deriving Show

data Map a@N b = Map a b deriving Show

main = do
    print (coerce $ one                       :: Age)
    print (coerce $ Age 1                     :: Int)
    print (coerce $ Baz (Bar (Age 1))         :: Foo)

    print (coerce (id::Bar->Bar) (Age 1)      :: Foo)
    print (coerce Baz (Age 1)                 :: Foo)
    print (coerce $ (Age 1, Foo (Age 1))      :: (Baz, Baz))

    print (coerce $ Map one one               :: Map Int Age)

  where one = 1 :: Int



