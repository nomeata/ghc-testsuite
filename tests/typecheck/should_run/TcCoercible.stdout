Age 1
1
Foo (Age 1)
Foo (Age 1)
Foo (Age 1)
(Baz (Bar (Age 1)),Baz (Bar (Age 1)))
Map 1 (Age 1)
First {getFirst = Just 1}
Last {getLast = Nothing}
